flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/IteratorSourceReader.java

```diff
@@ -71,40 +70,67 @@
      */
     @Nullable private SplitT currentSplit;
 
-    /** The remaining splits. Null means no splits have yet been assigned. */
-    @Nullable private Queue<SplitT> remainingSplits;
+    /** The remaining splits that were assigned but not yet processed. */
+    private Queue<SplitT> remainingSplits;
+
+    private boolean noMoreSplits;
 
     public IteratorSourceReader(SourceReaderContext context) {
         this.context = checkNotNull(context);
         this.availability = new CompletableFuture<>();
+        this.remainingSplits = new ArrayDeque<>();
     }
 
     // ------------------------------------------------------------------------
 
     @Override
     public void start() {
-        // request a split only if we did not get one during restore
-        if (remainingSplits == null) {
+        // request a split if we don't have one
+        if (remainingSplits.isEmpty()) {
             context.sendSplitRequest();
         }
     }
 
     @Override
     public InputStatus pollNext(ReaderOutput<E> output) {
-        if (iterator != null && iterator.hasNext()) {
-            output.collect(iterator.next());
+        if (iterator != null) {
+            if (iterator.hasNext()) {
+                output.collect(iterator.next());
+                return InputStatus.MORE_AVAILABLE;
+            } else {
+                finishSplit();
+            }
+        }
+
+        return tryMoveToNextSplit();
+    }
+
+    private void finishSplit() {
+        iterator = null;
+        currentSplit = null;
+
+        // request another split if no other is left
+        // we do this only here in the finishSplit part to avoid requesting a split
+        // whenever the reader is polled and doesn't currently have a split
+        if (remainingSplits.isEmpty() && !noMoreSplits) {
```

Comment:
If we add it there, then we will send a RequestSplit message every time the `pollNext()` message is called when the list is empty. That we need to avoid.

Having it here means we only check to request splits at start and when a split is finished. No duplicate requests due to polling the `pollNext()` method.

Reply:
I overlooked the comment apparently 😅 .